File: bindings/ruby/amdb.rb
# AmDb Ruby绑定
# 使用FFI调用C API

require 'ffi'

module AmDb
  extend FFI::Library
  
  # 加载C库
  ffi_lib './libamdb.so'
  
  # 定义C结构
  class AmdbResult < FFI::Struct
    layout :status, :int,
           :error_msg, :pointer,
           :data, :pointer,
           :data_len, :size_t
  end
  
  # 定义C函数
  attach_function :amdb_init, [:string, :pointer], :int
  attach_function :amdb_close, [:pointer], :int
  attach_function :amdb_put, [:pointer, :pointer, :size_t, :pointer, :size_t, :pointer], :int
  attach_function :amdb_get, [:pointer, :pointer, :size_t, :uint32, :pointer], :int
  attach_function :amdb_delete, [:pointer, :pointer, :size_t], :int
  attach_function :amdb_get_root_hash, [:pointer, :pointer], :int
  attach_function :amdb_free_result, [:pointer], :void
  attach_function :amdb_error_string, [:int], :string
  
  class Database
    def initialize(data_dir)
      @handle = FFI::MemoryPointer.new(:pointer)
      status = AmDb.amdb_init(data_dir, @handle)
      if status != 0
        raise "Failed to initialize database: #{AmDb.amdb_error_string(status)}"
      end
      @handle_value = @handle.read_pointer
    end
    
    def close
      if @handle_value
        AmDb.amdb_close(@handle_value)
        @handle_value = nil
      end
    end
    
    def put(key, value)
      key_bytes = key.is_a?(String) ? key.bytes : key
      value_bytes = value.is_a?(String) ? value.bytes : value
      
      root_hash = FFI::MemoryPointer.new(:uint8, 32)
      
      key_ptr = FFI::MemoryPointer.new(:uint8, key_bytes.length)
      key_ptr.write_bytes(key_bytes.pack('C*'))
      
      value_ptr = FFI::MemoryPointer.new(:uint8, value_bytes.length)
      value_ptr.write_bytes(value_bytes.pack('C*'))
      
      status = AmDb.amdb_put(
        @handle_value,
        key_ptr,
        key_bytes.length,
        value_ptr,
        value_bytes.length,
        root_hash
      )
      
      if status != 0
        raise "Put failed: #{AmDb.amdb_error_string(status)}"
      end
      
      root_hash.read_array_of_uint8(32)
    end
    
    def get(key, version = 0)
      key_bytes = key.is_a?(String) ? key.bytes : key
      key_ptr = FFI::MemoryPointer.new(:uint8, key_bytes.length)
      key_ptr.write_bytes(key_bytes.pack('C*'))
      
      result = AmdbResult.new
      status = AmDb.amdb_get(@handle_value, key_ptr, key_bytes.length, version, result)
      
      begin
        if status == -2  # AMDB_NOT_FOUND
          return nil
        end
        
        if status != 0
          raise "Get failed: #{AmDb.amdb_error_string(status)}"
        end
        
        if result[:data_len] > 0
          data_ptr = result[:data]
          data = data_ptr.read_bytes(result[:data_len])
          return data
        end
        
        nil
      ensure
        AmDb.amdb_free_result(result)
      end
    end
    
    def delete(key)
      key_bytes = key.is_a?(String) ? key.bytes : key
      key_ptr = FFI::MemoryPointer.new(:uint8, key_bytes.length)
      key_ptr.write_bytes(key_bytes.pack('C*'))
      
      status = AmDb.amdb_delete(@handle_value, key_ptr, key_bytes.length)
      if status != 0
        raise "Delete failed: #{AmDb.amdb_error_string(status)}"
      end
    end
    
    def get_root_hash
      root_hash = FFI::MemoryPointer.new(:uint8, 32)
      status = AmDb.amdb_get_root_hash(@handle_value, root_hash)
      if status != 0
        raise "Get root hash failed: #{AmDb.amdb_error_string(status)}"
      end
      root_hash.read_array_of_uint8(32)
    end
  end
end

File: bindings/nodejs/amdb.js
/**
 * AmDb Node.js绑定
 * 使用node-ffi或node-addon-api
 */

const ffi = require('ffi-napi');
const ref = require('ref-napi');
const Struct = require('ref-struct-di')(ref);

// 定义C类型
const uint8_t = ref.types.uint8;
const uint32_t = ref.types.uint32;
const size_t = ref.types.size_t;
const voidPtr = ref.refType(ref.types.void);

// 定义结构体
const AmdbResult = Struct({
    status: 'int',
    error_msg: 'string',
    data: voidPtr,
    data_len: size_t
});

// 加载C库
const amdb = ffi.Library('./libamdb', {
    'amdb_init': ['int', ['string', 'pointer']],
    'amdb_close': ['int', ['pointer']],
    'amdb_put': ['int', ['pointer', 'pointer', size_t, 'pointer', size_t, 'pointer']],
    'amdb_get': ['int', ['pointer', 'pointer', size_t, uint32_t, 'pointer']],
    'amdb_delete': ['int', ['pointer', 'pointer', size_t]],
    'amdb_batch_put': ['int', ['pointer', 'pointer', 'pointer', 'pointer', 'pointer', size_t, 'pointer']],
    'amdb_get_root_hash': ['int', ['pointer', 'pointer']],
    'amdb_free_result': ['void', ['pointer']],
    'amdb_error_string': ['string', ['int']]
});

class Database {
    constructor(dataDir) {
        this.handle = ref.alloc('pointer');
        const status = amdb.amdb_init(dataDir, this.handle);
        if (status !== 0) {
            throw new Error(`Failed to initialize database: ${amdb.amdb_error_string(status)}`);
        }
    }

    close() {
        const status = amdb.amdb_close(this.handle.deref());
        if (status !== 0) {
            throw new Error(`Failed to close database: ${amdb.amdb_error_string(status)}`);
        }
    }

    put(key, value) {
        const keyBuf = Buffer.from(key);
        const valueBuf = Buffer.from(value);
        const rootHash = Buffer.alloc(32);
        
        const status = amdb.amdb_put(
            this.handle.deref(),
            keyBuf,
            keyBuf.length,
            valueBuf,
            valueBuf.length,
            rootHash
        );
        
        if (status !== 0) {
            throw new Error(`Put failed: ${amdb.amdb_error_string(status)}`);
        }
        
        return rootHash;
    }

    get(key, version = 0) {
        const keyBuf = Buffer.from(key);
        const result = new AmdbResult();
        
        const status = amdb.amdb_get(
            this.handle.deref(),
            keyBuf,
            keyBuf.length,
            version,
            result.ref()
        );
        
        if (status !== 0) {
            if (status === -2) { // AMDB_NOT_FOUND
                return null;
            }
            throw new Error(`Get failed: ${amdb.amdb_error_string(status)}`);
        }
        
        if (result.data_len > 0) {
            const data = ref.readPointer(result.data, 0, result.data_len);
            return Buffer.from(data);
        }
        
        amdb.amdb_free_result(result.ref());
        return null;
    }

    delete(key) {
        const keyBuf = Buffer.from(key);
        const status = amdb.amdb_delete(this.handle.deref(), keyBuf, keyBuf.length);
        if (status !== 0) {
            throw new Error(`Delete failed: ${amdb.amdb_error_string(status)}`);
        }
    }

    batchPut(items) {
        const count = items.length;
        const keys = items.map(item => Buffer.from(item.key));
        const values = items.map(item => Buffer.from(item.value));
        
        // 创建指针数组
        const keyPtrs = keys.map(k => k);
        const valuePtrs = values.map(v => v);
        const keyLens = keys.map(k => k.length);
        const valueLens = values.map(v => v.length);
        
        const rootHash = Buffer.alloc(32);
        const status = amdb.amdb_batch_put(
            this.handle.deref(),
            keyPtrs,
            keyLens,
            valuePtrs,
            valueLens,
            count,
            rootHash
        );
        
        if (status !== 0) {
            throw new Error(`Batch put failed: ${amdb.amdb_error_string(status)}`);
        }
        
        return rootHash;
    }

    getRootHash() {
        const rootHash = Buffer.alloc(32);
        const status = amdb.amdb_get_root_hash(this.handle.deref(), rootHash);
        if (status !== 0) {
            throw new Error(`Get root hash failed: ${amdb.amdb_error_string(status)}`);
        }
        return rootHash;
    }
}

module.exports = { Database };

